File: lattice-ts/src/wire.ts
/**
 * Lattice wire format (spec §9): NDJSON record types over HTTP, entity refs,
 * field-value shapes, and a streaming record decoder.
 *
 * Forward tolerance (§9.4.1): unknown record kinds and unknown error-scope
 * tags MUST be tolerated; they surface here as `UnknownRecord` / opaque scope
 * objects rather than throwing.
 */

// ---------------------------------------------------------------------------
// JSON

export type JsonValue = null | boolean | number | string | JsonValue[] | JsonObject;
export interface JsonObject {
  [key: string]: JsonValue;
}

// ---------------------------------------------------------------------------
// Refs and field values

/** A typed entity reference: `"Type:key"`. */
export type Ref = string;

/** A to-one edge value on the wire: `{"$ref":"User:9"}`. */
export interface RefValue {
  readonly $ref: Ref;
}

/** The payload of a paginated edge: `{"items":[{"$ref":...}],"next":...,"prev":...}`. */
export interface Page {
  readonly items: readonly RefValue[];
  readonly next: string | null;
  readonly prev: string | null;
  readonly total?: number;
}

/** A paginated edge value on the wire: `{"$page":{...}}`. */
export interface PageValue {
  readonly $page: Page;
}

/**
 * Any value stored under an entity's field key: a plain JSON scalar or value,
 * a to-one `{$ref}` edge, a paginated `{$page}` edge, or (for bounded
 * collections) a plain array of ref strings.
 */
export type FieldValue = JsonValue | RefValue | PageValue;

export class LatticeWireError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "LatticeWireError";
  }
}

/** Split a `"Type:key"` ref into its type and key. Keys may themselves contain `:`. */
export function parseRef(ref: Ref): [type: string, key: string] {
  const i = ref.indexOf(":");
  if (i <= 0 || i === ref.length - 1) {
    throw new LatticeWireError(`malformed entity ref: ${JSON.stringify(ref)}`);
  }
  return [ref.slice(0, i), ref.slice(i + 1)];
}

/** The `Type` part of a `"Type:key"` ref. */
export function refType(ref: Ref): string {
  return parseRef(ref)[0];
}

/** Loose structural test for a ref string. */
export function isRef(v: unknown): v is Ref {
  if (typeof v !== "string") return false;
  const i = v.indexOf(":");
  return i > 0 && i < v.length - 1;
}

function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/** Is this field value a to-one edge (`{$ref}`)? */
export function isRefValue(v: unknown): v is RefValue {
  return isPlainObject(v) && typeof v["$ref"] === "string";
}

/** Is this field value a paginated edge (`{$page}`)? */
export function isPageValue(v: unknown): v is PageValue {
  if (!isPlainObject(v)) return false;
  const page = v["$page"];
  return isPlainObject(page) && Array.isArray(page["items"]);
}

/** Is this field value a bounded collection (a plain array of ref strings)? */
export function isRefArray(v: unknown): v is Ref[] {
  return Array.isArray(v) && v.every((x) => isRef(x) || isRefValue(x));
}

// ---------------------------------------------------------------------------
// Records

/** First record of every successful data-slice response (§9.2). */
export interface ManifestRecord {
  readonly kind: "manifest";
  readonly query?: string;
  readonly mutation?: string;
  readonly plan?: string;
  readonly slice?: string;
  /** Result order per root name (per item key `"items"` for batch mutations). */
  readonly root: Readonly<Record<string, readonly Ref[]>>;
  readonly etag?: string;
  readonly batch?: { readonly atomicity?: string; readonly count?: number };
}

export interface EntityRecord {
  readonly kind: "entity";
  readonly id: Ref;
  readonly ver: string;
  readonly fields: Readonly<Record<string, FieldValue>>;
  /** Batch-mutation item correlation (§11.8). */
  readonly item?: string;
  /** Federation source (§18.4). */
  readonly src?: string;
}

/** The entity no longer exists; clients MUST evict it (§9.3). */
export interface TombstoneRecord {
  readonly kind: "tombstone";
  readonly id: Ref;
  readonly ver: string;
}

/** The entity exists but this context may not see it; NOT nonexistence (§9.3). */
export interface ElidedRecord {
  readonly kind: "elided";
  readonly id: Ref;
}

/** Cache-digest elision marker: the client's `(id, ver)` is still current (§10.4). */
export interface UnchangedRecord {
  readonly kind: "unchanged";
  readonly id: Ref;
  readonly ver: string;
}

/**
 * What failed, per §9.4.1. `Entity` scope has a bare-ref string shorthand;
 * other constructors serialize as tagged sums. Unknown `$tag`s must be
 * tolerated (treated as unscoped for display).
 */
export type ErrorScope =
  | Ref
  | { readonly $tag: "Field"; readonly entity: Ref; readonly field: string }
  | { readonly $tag: "Edge"; readonly entity: Ref; readonly field: string }
  | { readonly $tag: "Root"; readonly root: string }
  | { readonly $tag: "Item"; readonly item: string }
  | { readonly $tag: string; readonly [k: string]: JsonValue | undefined };

export interface ErrorRecord {
  readonly kind: "error";
  readonly scope?: ErrorScope;
  /** Protocol vocabulary, `lattice:` namespace (§9.4.2). */
  readonly code?: string;
  /** Declared domain error sum, `{$tag: "AlreadyFilled", ...}` (§9.4.2). */
  readonly error?: { readonly $tag: string; readonly [k: string]: JsonValue | undefined };
  readonly retryable: boolean;
  readonly message?: string;
  readonly item?: string;
}

/** Mirror of the purge set so client stores can mark cached queries stale (§11.3). */
export interface InvalidatedRecord {
  readonly kind: "invalidated";
  readonly keys: readonly string[];
  readonly item?: string;
}

export interface EndRecord {
  readonly kind: "end";
  /** Did the origin finish attempting the full plan (§9.4.4)? Orthogonal to errors. */
  readonly complete: boolean;
  readonly etag?: string;
}

/** The single record of a `slice=plan` response (§9.2). */
export interface PlanRecord {
  readonly kind: "plan";
  readonly query?: string;
  readonly plan?: string;
  readonly slices?: JsonValue;
}

export interface ProgressRecord {
  readonly kind: "progress";
  readonly [k: string]: JsonValue | undefined;
}

/** A record kind this client does not know. Kept, never fatal (§9.4.1). */
export interface UnknownRecord {
  readonly kind: "unknown";
  readonly rawKind: string;
  readonly record: JsonObject;
}

export type LatticeRecord =
  | ManifestRecord
  | EntityRecord
  | TombstoneRecord
  | ElidedRecord
  | UnchangedRecord
  | ErrorRecord
  | InvalidatedRecord
  | EndRecord
  | PlanRecord
  | ProgressRecord
  | UnknownRecord;

// ---------------------------------------------------------------------------
// Record classification

function asUnknown(kind: string, record: JsonObject): UnknownRecord {
  return { kind: "unknown", rawKind: kind, record };
}

/**
 * Classify one decoded NDJSON value into a typed record. Records of unknown
 * kind, and records of a known kind that are missing required members, are
 * returned as `UnknownRecord` rather than thrown (forward tolerance).
 */
export function parseRecord(value: unknown): LatticeRecord {
  if (!isPlainObject(value)) {
    throw new LatticeWireError(`wire record is not a JSON object: ${JSON.stringify(value)}`);
  }
  const rec = value as JsonObject;
  const kind = typeof rec["kind"] === "string" ? (rec["kind"] as string) : "";
  switch (kind) {
    case "manifest": {
      const root = rec["root"];
      return {
        kind: "manifest",
        ...(typeof rec["query"] === "string" ? { query: rec["query"] } : {}),
        ...(typeof rec["mutation"] === "string" ? { mutation: rec["mutation"] } : {}),
        ...(typeof rec["plan"] === "string" ? { plan: rec["plan"] } : {}),
        ...(typeof rec["slice"] === "string" ? { slice: rec["slice"] } : {}),
        root: isPlainObject(root) ? (root as Record<string, Ref[]>) : {},
        ...(typeof rec["etag"] === "string" ? { etag: rec["etag"] } : {}),
        ...(isPlainObject(rec["batch"]) ? { batch: rec["batch"] as ManifestRecord["batch"] } : {}),
      };
    }
    case "entity": {
      if (typeof rec["id"] !== "string" || typeof rec["ver"] !== "string") return asUnknown(kind, rec);
      return {
        kind: "entity",
        id: rec["id"],
        ver: rec["ver"],
        fields: isPlainObject(rec["fields"]) ? (rec["fields"] as Record<string, FieldValue>) : {},
        ...(typeof rec["item"] === "string" ? { item: rec["item"] } : {}),
        ...(typeof rec["src"] === "string" ? { src: rec["src"] } : {}),
      };
    }
    case "tombstone": {
      if (typeof rec["id"] !== "string") return asUnknown(kind, rec);
      return { kind: "tombstone", id: rec["id"], ver: typeof rec["ver"] === "string" ? rec["ver"] : "" };
    }
    case "elided": {
      if (typeof rec["id"] !== "string") return asUnknown(kind, rec);
      return { kind: "elided", id: rec["id"] };
    }
    case "unchanged": {
      if (typeof rec["id"] !== "string") return asUnknown(kind, rec);
      return { kind: "unchanged", id: rec["id"], ver: typeof rec["ver"] === "string" ? rec["ver"] : "" };
    }
    case "error": {
      return {
        kind: "error",
        ...(rec["scope"] !== undefined ? { scope: rec["scope"] as ErrorScope } : {}),
        ...(typeof rec["code"] === "string" ? { code: rec["code"] } : {}),
        ...(isPlainObject(rec["error"]) ? { error: rec["error"] as ErrorRecord["error"] } : {}),
        retryable: rec["retryable"] === true,
        ...(typeof rec["message"] === "string" ? { message: rec["message"] } : {}),
        ...(typeof rec["item"] === "string" ? { item: rec["item"] } : {}),
      };
    }
    case "invalidated": {
      const keys = rec["keys"];
      return {
        kind: "invalidated",
        keys: Array.isArray(keys) ? keys.filter((k): k is string => typeof k === "string") : [],
        ...(typeof rec["item"] === "string" ? { item: rec["item"] } : {}),
      };
    }
    case "end":
      return {
        kind: "end",
        complete: rec["complete"] === true,
        ...(typeof rec["etag"] === "string" ? { etag: rec["etag"] } : {}),
      };
    case "plan":
      return {
        kind: "plan",
        ...(typeof rec["query"] === "string" ? { query: rec["query"] } : {}),
        ...(typeof rec["plan"] === "string" ? { plan: rec["plan"] } : {}),
        ...(rec["slices"] !== undefined ? { slices: rec["slices"] } : {}),
      };
    case "progress":
      return rec as unknown as ProgressRecord;
    default:
      return asUnknown(kind, rec);
  }
}

// ---------------------------------------------------------------------------
// Stream decoding

function decodeLine(line: string, strict: boolean): LatticeRecord | undefined {
  const trimmed = line.endsWith("\r") ? line.slice(0, -1) : line;
  if (trimmed.trim() === "") return undefined;
  let json: unknown;
  try {
    json = JSON.parse(trimmed);
  } catch (e) {
    if (strict) {
      throw new LatticeWireError(`malformed NDJSON line: ${trimmed.slice(0, 120)}`);
    }
    return undefined; // trailing partial line of a truncated stream
  }
  return parseRecord(json);
}

/**
 * Stream-decode an NDJSON entity stream from a fetch `Response`.
 *
 * Complete lines must parse (a malformed complete line throws
 * `LatticeWireError`); a trailing unterminated partial line — the signature of
 * a truncated transfer — is tolerated and dropped (the missing `end` record
 * already marks the response incomplete).
 */
export async function* readRecords(response: Response): AsyncGenerator<LatticeRecord, void, unknown> {
  const body = response.body;
  if (!body) {
    // Environments that buffered the body (or an empty response).
    const text = await response.text();
    yield* recordsOfText(text);
    return;
  }
  const reader = body.getReader();
  const decoder = new TextDecoder("utf-8");
  let buf = "";
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });
      let nl: number;
      while ((nl = buf.indexOf("\n")) >= 0) {
        const line = buf.slice(0, nl);
        buf = buf.slice(nl + 1);
        const rec = decodeLine(line, true);
        if (rec) yield rec;
      }
    }
    buf += decoder.decode();
    const tail = decodeLine(buf, false);
    if (tail) yield tail;
  } finally {
    reader.releaseLock();
  }
}

/** Decode records from an already-buffered NDJSON string. */
export function* recordsOfText(text: string): Generator<LatticeRecord, void, unknown> {
  const lines = text.split("\n");
  for (let i = 0; i < lines.length; i++) {
    const last = i === lines.length - 1;
    const rec = decodeLine(lines[i] ?? "", !last);
    if (rec) yield rec;
  }
}

// ---------------------------------------------------------------------------
// Denormalized data shape (what `client.denormalize` produces)

/** A denormalized entity: its ref plus the selected fields as plain properties. */
export type EntityTree = { readonly __ref: Ref } & { [field: string]: unknown };

/** A denormalized paginated edge. */
export interface PageTree {
  readonly items: EntityTree[];
  readonly next: string | null;
  readonly prev: string | null;
  readonly total?: number;
}

/** Default shape of a denormalized query result: one entry per root field. */
export type QueryData = { [root: string]: EntityTree[] | undefined };

// ---------------------------------------------------------------------------
// Header / URL helpers shared by client and server-ish code

/** Well-known Lattice header names. */
export const HEADERS = {
  plan: "lattice-plan",
  schema: "lattice-schema",
  snapshot: "lattice-snapshot",
  outcome: "lattice-outcome",
  queryName: "lattice-query-name",
  surrogateKey: "surrogate-key",
  vcAuth: "x-vc-auth",
  idempotencyKey: "idempotency-key",
} as const;

/** The query media type used by the introduction rungs (§6.3/§6.4). */
export const QUERY_MEDIA_TYPE = "application/x-lattice-query";

/**
 * URL parameter names reserved by the protocol; variable names must not
 * collide with these.
 */
export const RESERVED_PARAMS: Record<string, true> = {
  p: true,
  slice: true,
  vc: true,
  project: true,
  live: true,
  d: true,
  dv: true,
  intent: true,
};

File: wireform-lattice/src/Lattice/Wire.hs
{- | The Lattice wire format (spec §9): NDJSON entity-stream records, the
manifest, scoped errors, and the page value convention, plus the protocol
header names shared by server and client.

JSON shapes are pinned here as the cross-language contract (the TypeScript
client mirrors them):

@
{"kind":"manifest","query":"8f2c…","plan":"pl_…","slice":"ctx","root":{"feed":["Post:17"]},"etag":"m:…"}
{"kind":"entity","id":"Post:17","ver":"e41","fields":{"title":"…","author":{"$ref":"User:9"}}}
{"kind":"tombstone","id":"Post:17","ver":"t:99"}
{"kind":"elided","id":"Post:17"}
{"kind":"unchanged","id":"Post:17","ver":"e41"}
{"kind":"error","scope":"Human:1002","code":"lattice:loader-timeout","retryable":true}
{"kind":"invalidated","keys":["Post:17","feed:123"]}
{"kind":"end","complete":true,"etag":"m:…"}
{"kind":"plan","query":"8f2c…","plan":"pl_…","slices":{"pub":false,"ctx":{"claims":["org"],"roots":["feed"]},"priv":false}}
@

Unknown record kinds and unknown scope tags decode to their tolerant
constructors ('RUnknown', 'ScopeUnknown'); clients MUST NOT fail on them
(§9.4.1).
-}
module Lattice.Wire (
  -- * Records
  Record (..),
  Manifest (..),
  BatchInfo (..),
  EntityRecord (..),
  ErrorRecord (..),
  Scope (..),
  EndRecord (..),
  PlanRecord (..),
  SliceInfo (..),

  -- * Page values
  PageValue (..),
  pageToJSON,
  pageFromJSON,
  refValue,

  -- * NDJSON framing
  encodeRecords,
  encodeRecord,
  decodeRecords,

  -- * Surrogate keys
  SurrogateKey,
  entityKeyOf,
  collectionKey,
  planKey,

  -- * Protocol header names
  hLatticePlan,
  hLatticeSchema,
  hLatticeSnapshot,
  hLatticeOutcome,
  hLatticeQueryName,
  hSurrogateKey,
  hIdempotencyKey,
  hIdempotencyReplayed,
  hVcAuth,
  hXHave,
  hXHaveDigest,
  hLatticeQuerySig,
  hLatticeClient,

  -- * Media types
  queryMediaType,
) where

import Data.Aeson (FromJSON (..), ToJSON (..), (.:), (.:?), (.=))
import Data.Aeson qualified as A
import Data.Aeson.Key qualified as AK
import Data.Aeson.KeyMap qualified as KM
import Data.Aeson.Types qualified as A
import Data.ByteString (ByteString)
import Data.ByteString.Char8 qualified as BS8
import Data.ByteString.Lazy qualified as BL
import Data.CaseInsensitive (CI)
import Data.Map.Strict (Map)
import Data.Map.Strict qualified as Map
import Data.Maybe (catMaybes, mapMaybe)
import Data.Text (Text)
import Lattice.Types


-- ---------------------------------------------------------------------------
-- Records
-- ---------------------------------------------------------------------------

data Record
  = RManifest Manifest
  | REntity EntityRecord
  | RTombstone Ref Text (Maybe Text)
  -- ^ id, ver, item key.
  | RElided Ref
  | RUnchanged Ref Text
  | RError ErrorRecord
  | RInvalidated [SurrogateKey] (Maybe Text)
  -- ^ keys, item key.
  | REnd EndRecord
  | RPlan PlanRecord
  | RReauth
  | -- | Tolerated unknown kind: kept verbatim for forward compatibility.
    RUnknown Text A.Object
  deriving stock (Eq, Show)


data Manifest = Manifest
  { mQuery :: Maybe Text
  -- ^ Query hash; present on query responses.
  , mMutation :: Maybe MutationName
  -- ^ Present on mutation responses.
  , mPlan :: Maybe Text
  , mSlice :: Maybe SliceName
  , mRoot :: Map Text [Ref]
  -- ^ Result order per root name (or @result@ / @items@ for mutations).
  , mEtag :: Text
  , mBatch :: Maybe BatchInfo
  }
  deriving stock (Eq, Show)


data BatchInfo = BatchInfo
  { biAtomicity :: Text
  -- ^ @"all-or-nothing"@ | @"best-effort"@.
  , biCount :: Int
  }
  deriving stock (Eq, Show)


data EntityRecord = EntityRecord
  { erId :: Ref
  , erVer :: Text
  , erFields :: Map Text A.Value
  -- ^ Keyed by canonical field form, e.g. @avatarUrl(size:48)@.
  , erItem :: Maybe Text
  -- ^ Batch item correlation key.
  , erSrc :: Maybe Text
  -- ^ Federation source tag (§18.4).
  }
  deriving stock (Eq, Show)


-- | The protocol-defined open Scope sum (§9.4.1).
data Scope
  = ScopeEntity Ref
  | ScopeField Ref FieldName
  | ScopeEdge Ref FieldName
  | ScopeRoot RootName
  | ScopeItem Text
  | -- | Unrecognized @$tag@, kept verbatim; treat as unscoped for display.
    ScopeUnknown Text A.Value
  deriving stock (Eq, Show)


data ErrorRecord = ErrorRecord
  { errScope :: Maybe Scope
  -- ^ 'Nothing' for a fatal, unscoped stream error.
  , errCode :: Maybe Text
  -- ^ Protocol vocabulary, @lattice:*@.
  , errDomain :: Maybe A.Value
  -- ^ Declared domain error sum value, @{"$tag":…}@.
  , errRetryable :: Bool
  , errMessage :: Maybe Text
  }
  deriving stock (Eq, Show)


data EndRecord = EndRecord
  { endComplete :: Bool
  , endEtag :: Maybe Text
  }
  deriving stock (Eq, Show)


-- | The @slice=plan@ response record (§9.2).
data PlanRecord = PlanRecord
  { prQuery :: Text
  , prPlan :: Text
  , prSlices :: Map SliceName SliceInfo
  -- ^ Only data slices appear; a missing key means the slice is empty.
  }
  deriving stock (Eq, Show)


data SliceInfo = SliceInfo
  { siClaims :: [ClaimName]
  , siRoots :: [RootName]
  }
  deriving stock (Eq, Show)


-- ---------------------------------------------------------------------------
-- JSON instances
-- ---------------------------------------------------------------------------

instance ToJSON Record where
  toJSON = \case
    RManifest m -> obj ("manifest" :: Text) (manifestPairs m)
    REntity e ->
      obj
        "entity"
        ( [ "id" .= renderRef (erId e)
          , "ver" .= erVer e
          , "fields" .= A.object [AK.fromText k .= v | (k, v) <- Map.toList (erFields e)]
          ]
            <> catMaybes
              [ ("item" .=) <$> erItem e
              , ("src" .=) <$> erSrc e
              ]
        )
    RTombstone r v item ->
      obj "tombstone" (["id" .= renderRef r, "ver" .= v] <> catMaybes [("item" .=) <$> item])
    RElided r -> obj "elided" ["id" .= renderRef r]
    RUnchanged r v -> obj "unchanged" ["id" .= renderRef r, "ver" .= v]
    RError e ->
      obj
        "error"
        ( catMaybes
            [ ("scope" .=) <$> errScope e
            , ("code" .=) <$> errCode e
            , ("error" .=) <$> errDomain e
            , Just ("retryable" .= errRetryable e)
            , ("message" .=) <$> errMessage e
            ]
        )
    RInvalidated keys item ->
      obj "invalidated" (["keys" .= keys] <> catMaybes [("item" .=) <$> item])
    REnd e ->
      obj "end" (["complete" .= endComplete e] <> catMaybes [("etag" .=) <$> endEtag e])
    RPlan p ->
      obj
        "plan"
        [ "query" .= prQuery p
        , "plan" .= prPlan p
        , "slices"
            .= A.object
              [ AK.fromText (renderSlice s)
                  .= maybe (A.Bool False) sliceInfoJSON (Map.lookup s (prSlices p))
              | s <- [SlicePub, SliceCtx, SlicePriv]
              ]
        ]
    RReauth -> obj "reauth" []
    RUnknown k o -> A.Object (KM.insert "kind" (A.String k) o)
    where
      obj :: Text -> [A.Pair] -> A.Value
      obj k ps = A.object (("kind" .= k) : ps)
      sliceInfoJSON (SliceInfo cs rs) =
        A.object
          [ "claims" .= map unClaimName cs
          , "roots" .= map unRootName rs
          ]
      manifestPairs m =
        catMaybes
          [ ("query" .=) <$> mQuery m
          , ("mutation" .=) . unMutationName <$> mMutation m
          , ("plan" .=) <$> mPlan m
          , ("slice" .=) . renderSlice <$> mSlice m
          , ("batch" .=) <$> mBatch m
          ]
          <> [ "root" .= A.object [AK.fromText k .= map renderRef rs | (k, rs) <- Map.toList (mRoot m)]
             , "etag" .= mEtag m
             ]


instance FromJSON Record where
  parseJSON = A.withObject "Record" $ \o -> do
    kind <- o .: "kind"
    case kind :: Text of
      "manifest" -> do
        mQuery <- o .:? "query"
        mMutationT <- o .:? "mutation"
        mPlan <- o .:? "plan"
        mSliceT <- o .:? "slice"
        rootObj <- o .:? "root" A..!= KM.empty
        mEtag <- o .: "etag"
        mBatch <- o .:? "batch"
        mRoot <-
          Map.fromList
            <$> traverse
              (\(k, v) -> (AK.toText k,) <$> parseRefList v)
              (KM.toList rootObj)
        let mMutation = MutationName <$> mMutationT
            mSlice = mSliceT >>= parseSlice
        pure (RManifest Manifest {..})
      "entity" -> do
        r <- refField o
        erVer <- o .: "ver"
        fieldsObj <- o .: "fields"
        erItem <- o .:? "item"
        erSrc <- o .:? "src"
        let erFields = Map.fromList [(AK.toText k, v) | (k, v) <- KM.toList fieldsObj]
        pure (REntity EntityRecord {erId = r, ..})
      "tombstone" -> RTombstone <$> refField o <*> o .: "ver" <*> o .:? "item"
      "elided" -> RElided <$> refField o
      "unchanged" -> RUnchanged <$> refField o <*> o .: "ver"
      "error" ->
        fmap RError $
          ErrorRecord
            <$> o .:? "scope"
            <*> o .:? "code"
            <*> o .:? "error"
            <*> o .:? "retryable" A..!= False
            <*> o .:? "message"
      "invalidated" -> RInvalidated <$> o .: "keys" <*> o .:? "item"
      "end" -> fmap REnd $ EndRecord <$> o .: "complete" <*> o .:? "etag"
      "plan" -> do
        prQuery <- o .: "query"
        prPlan <- o .: "plan"
        slicesObj <- o .: "slices"
        prSlices <-
          Map.fromList . catMaybes
            <$> traverse
              ( \(k, v) -> case parseSlice (AK.toText k) of
                  Nothing -> pure Nothing
                  Just s -> case v of
                    A.Bool False -> pure Nothing
                    _ -> do
                      info <- parseSliceInfo v
                      pure (Just (s, info))
              )
              (KM.toList slicesObj)
        pure (RPlan PlanRecord {..})
      "reauth" -> pure RReauth
      other -> pure (RUnknown other (KM.delete "kind" o))
    where
      refField o = do
        t <- o .: "id"
        maybe (fail "bad ref") pure (parseRef t)
      parseRefList = A.withArray "refs" $ \v ->
        traverse (A.withText "ref" (maybe (fail "bad ref") pure . parseRef)) (foldr (:) [] v)
      parseSliceInfo = A.withObject "SliceInfo" $ \s ->
        SliceInfo
          <$> (map ClaimName <$> s .:? "claims" A..!= [])
          <*> (map RootName <$> s .:? "roots" A..!= [])


instance ToJSON BatchInfo where
  toJSON (BatchInfo a c) = A.object ["atomicity" .= a, "count" .= c]


instance FromJSON BatchInfo where
  parseJSON = A.withObject "BatchInfo" $ \o ->
    BatchInfo <$> o .: "atomicity" <*> o .: "count"


{- | 'ScopeEntity' uses the bare-ref shorthand; every other constructor is a
tagged object (§9.4.1).
-}
instance ToJSON Scope where
  toJSON = \case
    ScopeEntity r -> A.String (renderRef r)
    ScopeField r f ->
      A.object ["$tag" .= ("Field" :: Text), "entity" .= renderRef r, "field" .= unFieldName f]
    ScopeEdge r f ->
      A.object ["$tag" .= ("Edge" :: Text), "entity" .= renderRef r, "field" .= unFieldName f]
    ScopeRoot n -> A.object ["$tag" .= ("Root" :: Text), "root" .= unRootName n]
    ScopeItem k -> A.object ["$tag" .= ("Item" :: Text), "item" .= k]
    ScopeUnknown tag v -> case v of
      A.Object o -> A.Object (KM.insert "$tag" (A.String tag) o)
      _ -> A.object ["$tag" .= tag, "value" .= v]


instance FromJSON Scope where
  parseJSON v = case v of
    A.String t -> maybe (fail "bad scope ref") (pure . ScopeEntity) (parseRef t)
    A.Object o -> do
      tag <- o .: "$tag"
      case tag :: Text of
        "Entity" -> ScopeEntity <$> refOf o "entity"
        "Field" -> ScopeField <$> refOf o "entity" <*> (FieldName <$> o .: "field")
        "Edge" -> ScopeEdge <$> refOf o "entity" <*> (FieldName <$> o .: "field")
        "Root" -> ScopeRoot . RootName <$> o .: "root"
        "Item" -> ScopeItem <$> o .: "item"
        other -> pure (ScopeUnknown other (A.Object (KM.delete "$tag" o)))
    _ -> fail "scope must be a string or object"
    where
      refOf o k = do
        t <- o .: k
        maybe (fail "bad ref") pure (parseRef t)


-- ---------------------------------------------------------------------------
-- Page values
-- ---------------------------------------------------------------------------

{- | The wire form of one paginated edge occurrence (§3.6):
@{"$page":{"items":[…],"next":…,"prev":…,"total":…}}@. Bounded collections
are plain ref-string arrays and never use this wrapper.
-}
data PageValue = PageValue
  { pvItems :: [Ref]
  , pvNext :: Maybe Text
  , pvPrev :: Maybe Text
  , pvTotal :: Maybe Int
  }
  deriving stock (Eq, Show)


pageToJSON :: PageValue -> A.Value
pageToJSON PageValue {..} =
  A.object
    [ "$page"
        .= A.object
          ( [ "items" .= map (\r -> A.object ["$ref" .= renderRef r]) pvItems
            , "next" .= pvNext
            , "prev" .= pvPrev
            ]
              <> maybe [] (\t -> ["total" .= t]) pvTotal
          )
    ]


pageFromJSON :: A.Value -> Maybe PageValue
pageFromJSON v = A.parseMaybe parser v
  where
    parser = A.withObject "$page" $ \o -> do
      p <- o .: "$page"
      flip (A.withObject "page body") p $ \b -> do
        items <- b .: "items"
        refs <- traverse itemRef items
        PageValue refs <$> b .:? "next" <*> b .:? "prev" <*> b .:? "total"
    itemRef = A.withObject "$ref" $ \o -> do
      t <- o .: "$ref"
      maybe (fail "bad ref") pure (parseRef t)


-- | The wire form of a to-one edge occurrence: @{"$ref":"Type:key"}@.
refValue :: Ref -> A.Value
refValue r = A.object ["$ref" .= renderRef r]


-- ---------------------------------------------------------------------------
-- NDJSON framing
-- ---------------------------------------------------------------------------

encodeRecord :: Record -> ByteString
encodeRecord = BL.toStrict . A.encode


-- | One record per line, each line newline-terminated.
encodeRecords :: [Record] -> ByteString
encodeRecords rs = BS8.concat [encodeRecord r <> "\n" | r <- rs]


{- | Decode an NDJSON body. Blank lines are skipped; undecodable lines are
returned as 'Left' with the offending line so callers can choose severity.
-}
decodeRecords :: ByteString -> [Either ByteString Record]
decodeRecords body =
  mapMaybe decodeLine (BS8.lines body)
  where
    decodeLine ln
      | BS8.all (== ' ') ln = Nothing
      | otherwise = Just $ case A.eitherDecodeStrict ln of
          Left _ -> Left ln
          Right r -> Right r


-- ---------------------------------------------------------------------------
-- Surrogate keys
-- ---------------------------------------------------------------------------

-- | A cache tag (§10.5): @Type:id@, @{collection}:{grouping}@, or @plan:{planId}@.
type SurrogateKey = Text


entityKeyOf :: Ref -> SurrogateKey
entityKeyOf = renderRef


collectionKey :: CollectionName -> [Text] -> SurrogateKey
collectionKey (CollectionName n) groupVals = n <> ":" <> commaJoin groupVals
  where
    commaJoin [] = ""
    commaJoin [x] = x
    commaJoin (x : xs) = x <> "," <> commaJoin xs


planKey :: Text -> SurrogateKey
planKey planId = "plan:" <> planId


-- ---------------------------------------------------------------------------
-- Headers and media types
-- ---------------------------------------------------------------------------

hLatticePlan, hLatticeSchema, hLatticeSnapshot, hLatticeOutcome :: CI ByteString
hLatticePlan = "Lattice-Plan"
hLatticeSchema = "Lattice-Schema"
hLatticeSnapshot = "Lattice-Snapshot"
hLatticeOutcome = "Lattice-Outcome"


hLatticeQueryName, hSurrogateKey, hIdempotencyKey, hIdempotencyReplayed :: CI ByteString
hLatticeQueryName = "Lattice-Query-Name"
hSurrogateKey = "Surrogate-Key"
hIdempotencyKey = "Idempotency-Key"
hIdempotencyReplayed = "Idempotency-Replayed"


hVcAuth, hXHave, hXHaveDigest, hLatticeQuerySig, hLatticeClient :: CI ByteString
hVcAuth = "X-Vc-Auth"
hXHave = "X-Have"
hXHaveDigest = "X-Have-Digest"
hLatticeQuerySig = "X-Lattice-Query-Sig"
hLatticeClient = "Lattice-Client"


queryMediaType :: ByteString
queryMediaType = "application/x-lattice-query"
